bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,34 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items == null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
+		Set<Item> items = null;
+		synchronized (contextMap) {
+			items = contextMap.get(context);
+			if (items != null) {
+				contextMap.remove(context);
+			}
+		}
+		if (items != null) {
+			for (Item item : items) {
 				// we remove all binding configurations for all items
 				bindingConfigs.remove(item.getName());
 				notifyListeners(item);
 			}
-			contextMap.remove(context);
 		}
 	}
 	
```

Comment:
Thanks for taking another look.  I also don't know if there are side effects in changing the ordering, but what I do know is that now, if multiple threads are calling `removeConfigurations` with the same `context`, as I suspect is what is causing to the NPEs, only the first will have any effect, and the others will essentially be no-ops.  Unless there is an intervening `processBindingConfiguration`, in which case the next thread to call `removeConfigurations` will be iterating over a completely different `Set` from the previous thread to enter `removeConfigurations`.  In other words, I don't see how multiple threads can step on each other when contending for the `contextMap` after these changes.

There are most likely higher-level coordination issues when files are reloaded, but I believe (and tell me if I'm wrong) that at least this code, in isolation, appears threadsafe.